// ngx-logger/esm5/lib/testing/logger.service.mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NGXLoggerMock = /** @class */ (function () {
    function NGXLoggerMock() {
    }
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    NGXLoggerMock.prototype.trace = function (message) {
        var additional = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            additional[_i - 1] = arguments[_i];
        }
    };
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    NGXLoggerMock.prototype.debug = function (message) {
        var additional = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            additional[_i - 1] = arguments[_i];
        }
    };
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    NGXLoggerMock.prototype.info = function (message) {
        var additional = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            additional[_i - 1] = arguments[_i];
        }
    };
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    NGXLoggerMock.prototype.log = function (message) {
        var additional = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            additional[_i - 1] = arguments[_i];
        }
    };
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    NGXLoggerMock.prototype.warn = function (message) {
        var additional = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            additional[_i - 1] = arguments[_i];
        }
    };
    /**
     * @param {?} message
     * @param {...?} additional
     * @return {?}
     */
    NGXLoggerMock.prototype.error = function (message) {
        var additional = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            additional[_i - 1] = arguments[_i];
        }
    };
    /**
     * @param {?} config
     * @return {?}
     */
    NGXLoggerMock.prototype.updateConfig = function (config) {
    };
    return NGXLoggerMock;
}());
export { NGXLoggerMock };
